### DISCUSSION TITLE
JIT plugin format

### DISCUSSION THREAD

**Discussion by Photosounder** [free-audio/clap#317] [Ideas]
Last year on KVR there was [some discussion](https://www.kvraudio.com/forum/viewtopic.php?p=8469401#p8469401) about the possibility of a CLAP JIT format. The idea was essentially that the idea of making one binary for each architecture and platform had run its course (it's inconvenient for the developer and terrible for futureproofing, as we've seen with old unupdated 32-bit plugins tragically fading into oblivion) and that we need a much more 2020s option of having one file that can potentially work on any platform, any OS or CPU architecture (even virtual ones like WebAssembly), even platforms that don't yet exist. I was told back then that this was under consideration, but I see nothing about this. Has this idea been dropped?

**Comment by luciascarlet** [comment]
I agree this is a good idea; it might not be suitable for every plugin necessarily, but it would be great to have. I feel like the CLAP format itself is generic enough that it might not really require any modifications to run under WASM, but in that case, there should be some generic and standardised method of providing a GUI that can run across all platforms. HTML and JavaScript seem like the natural solution for this.

**Comment by Photosounder** [comment]
HTML? Do many hosts render HTML and execute Javascript? I mean I don't hate the idea of having that option, but a more traditional and generic way to provide graphics is through a framebuffer that is updated by callback, that alone is good enough for most developers. But yes as long as your plugin sticks to portable code without strictly depending on platform-specific APIs then you could have one file for all platforms (which would be great for Linux as making Linux builds of plugins was never a popular choice despite the lack of obstacles).

**Comment by abique** [comment]
I've thought about this, but it is a lot of work.
You'd need to standardize the usage of WASI and also provide a canvas or cairo-like interface maybe, and maybe even some kind of windowing system abstraction.

For such a project to succeed, we'd need a full support library on the clap side. Making that one to be super reliable and well designed so big player will consider using it, isn't a part time job.

**Reply by Photosounder** [reply]
> provide a canvas or cairo-like interface maybe, and maybe even some kind of windowing system abstraction.

It's very simple guys, I don't know why you want to make it more complicated than it needs to be, it's like I said, you provide a framebuffer. That's what IPlug does, you instantiate a framebuffer by calling IPlug's `MakeGraphics()` like `IGraphics *pGraphics = MakeGraphics(this, width, height);`, then your framebuffer is in `pGraphics->GetDrawBitmap()->getBits()` so you draw to it in your control's `Draw()` function which is your drawing callback and that's it. A framebuffer is the common denominator, that's the first thing you need to provide for graphics, if you want to make plugin binaries that have a chance of still working 40 years from now then you want to rely on a simple framebuffer, anything else is just a bonus that might break well within the next couple of decades. My plugins SplineEQ and Spiral work entirely by updating IPlug's framebuffer. Maybe @olilarkin can provide insights about how this is to be handled.

> so big player will consider using it

No need for that as a first step, we can start with a bridge that hosts a CLAP-JIT plugin, then when everything is set in stone and does everything you want you can consider telling DAW creators to use this. The beauty of a bridge is that it doesn't even have to be in CLAP format (if it's up to me I'd write it using IPlug2), and once you update the bridge the update is available on every host.

> WASI

I didn't know about this, interesting, that does potentially simplify things a bit. Maybe I could even try something on my own, do we have any example plugins that don't rely on the ungodly abomination that is Qt?

Again, all you need from the host/bridge is:

- Call the plugin's initialisation callback, which defines things like parameters (you know, like controls, for automation) and the window's size, so after the callback returns the host/bridge takes care of creating the window and the parameters
- Call the processing callbacks, in the case of a bridge it's just a question of passing things from its own processing callbacks
- Call the param change callback, again for a bridge it's just a matter of passing things along
- Call the graphics callback, possibly with mouse and keyboard data as inputs (or these would be in separate callbacks), then when it returns display the framebuffer onto the window created by the host/bridge
- And of course call the terminating callback

**Reply by abique** [reply]
Even if it is just a bitmap buffer, you still need to provide mouse events, keyboard events, etc... So a windowing system abstraction more or less.

What do you do regarding plugins which requires hardware acceleration, shaders, ...?

**Reply by Photosounder** [reply]
> you still need to provide mouse events, keyboard events

I don't know why you want to make it sound like a problem, that's very simple, you just pass things along to the plugin's callback. Again, look at IPlug.

> What do you do regarding plugins which requires hardware acceleration, shaders, ...?

It's hard to make something universal and future-proof if you require something so dependent on platform-dependent APIs, isn't it? You could have specific callbacks for the API you want to target, but I don't think this would be worth the bother seeing as there's little advantage in using a universal plugin format if you want to target DirectX 12 or Metal 2. This is especially true now that OpenGL is kind of dead and there's no platform-independent standard to replace it (I don't think Vulkan, WebGL/WebGPU would count, at least not for the time being), if you want hardware accelerated graphics you're stuck in the ephemeral world of platform-specific APIs, as far as I know you can't make the same code be hardware accelerated on Windows, macOS, Linux and the web at the same time, so you'll always have to produce platform-specific binaries. If OpenGL was still viable I'd give you a different answer but here we are.

So this one falls firmly in the "not our problem" category because a universal plugin format can hardly be for you if you want to rely on such platform-specific features. A universal plugin format is for everyone else (definitely the majority of people who makes plugins) who writes portable code and wants to target every platform at once (Windows, macOS (God knows this one's a PITA to handle on your own mostly with notarising plugins), Linux, the web, any future platform that isn't yet on anyone's radar...) with everything working the same everywhere and not necessarily have to worry about their plugins becoming unsupported. So from this perspective there's no point in doing anything about people who need code that works differently on different platforms, you can release a .dll if you want DirectX and you can also release a JIT version if you can not use DirectX so it also works on Linux or whatever.

**Reply by abique** [reply]
If everything seems simple to you, then maybe start hacking something?

**Reply by Photosounder** [reply]
> If everything seems simple to you, then maybe start hacking something?

Yep, didn't think I could but the WASI thing might be just what I need, I'll look into that first, nothing else I can do until I can call functions from some compiled binary format.

Btw, does CLAP currently really not have a simple way to make a simple window and get mouse/keyboard events? Everyone is basically on their own when it comes to making a window?

**Reply by abique** [reply]
Yes, the plugin is talking directly with the windowing system.

**Reply by olilarkin** [reply]
PUGL seems like a good choice for the UI imo.

**Reply by baconpaul** [reply]
When thinking about UI choices, please please please remember to consider accessibility. The advent of accessible APIs in JUCE has made a set of synths available to a set of blind users who are locked out of most other technologies.

**Reply by baconpaul** [reply]
I think the reality of the innards of plugins means there's some things to think about with a JIT format which are tricky, but the two which come to top of mind are:

1. SIMD vectorization
2. Memory management and allocation management

I think most well running plugins have SIMD-vectorized innards and do a variety of tricks to make sure their memory is pre-allocated. These can be tough things to do if the abstraction is at the wrong level.

WASM did [some work](https://github.com/tc39/ecmascript_simd) on this it seems and, honestly, i think the best bet for a neutral cross platform close-to-metal representation of DSP code we have today is WASM. But I could also make a pretty strong case for LLVM byte code and then just require everyone to ship LLVM with their DAW (which is basically the strategy faust has taken). 

Faust is also an interesting case study here. They have a higher level language which projects to multiple back ends including platform assembly via LLVM. Definitely worth looking at.

Memory management is a harder issue. Things like pre-allocated voices and so on are tricky to engineer unless you use use C++ and project to a neutral back end (which is basically the WASM strategy).

Hope these thoughts are helpful

**Reply by abique** [reply]
Forcing hosts to ship LLVM is problematic, because they break the API every major release and it isn't possible to mix multiple versions in the same executable.

Also LLVM bytecode isn't stable.

I believe that WASM is a better target, and I believe that it is much more complicated than it seems. The challenge isn't to get a PoC, but to have everything working: font rendering, painting, interaction events, windowing, open file dialogs, use network sockets, ...

All that in a nice library which hosts will be interested to start using, but there also needs to be a nice package for the plugins.

**Reply by abique** [reply]
The promise of this feature is:
- you'll build your plugin once, and it'll work forever and on all platforms supported by WASM and WASI
- the performances will be comparable to native plugins
- you'll have less problems supporting a single WASM plugin than N native targets

I think that if one of those promise breaks, then the whole feature is dead.

**Reply by Photosounder** [reply]
I agree that WASI is the superior approach, although I don't know how stable it's gonna be, I think it's only a matter of time before it matures and stabilises (for instance they recently added some form of thread support). I'm reading a lot about it and it seems to have pretty much everything we need.

I'm gonna work in that direction but the way I envision it it will have little to do with CLAP, mostly due to my framebuffer philosophy and also so I don't have to worry about all of CLAP's features. We'll see after I get anything working if we can do anything to reconcile approaches or if anyone wants to adapt the idea for anything else but for me WASM plugins that can be hosted on an open source IPlug2 bridge plugin would satisfy what I'm looking for and I don't need much network effect (I don't need either DAW makers nor plugin markers to adopt my envisioned approach) beyond maybe a wide array of builds of the bridge plugin.

> have everything working: font rendering, painting, interaction events, windowing, open file dialogs, use network sockets, ...

That's why it will be better if I start with my own approach, my requirements are lower, my plugins already do font rendering and other rendering/GUI on their own, that's why I only need a framebuffer + mouse/keyboard inputs. We can worry about making a library full of sophisticated features after we have a proof of concept working. I don't work full time on plugins anymore though so it might take me some time.

**Reply by defiantnerd** [reply]
Regarding the topic UI:  The lack of a consistent UI system for C++ programmers is a result of the versatility of C++. There is not one fits it all solution. And today, things like HiDPI and platform restrictions make things even harder.

> It's very simple guys, I don't know why you want to make it more complicated than it needs to be, it's like I said, you provide a framebuffer.

And then the complicated things arrive: HiDPI, font rendering, anti-aliasing, bitmaps or vector formats, mouse events, keyboard events (some hosts don't let you do that), focus issues, rendering with 60Hz, accessibility, shaders... you can scale complexity at all dimensions.

A summary of current solutions can be found here: https://philippegroarke.com/posts/2018/c++_ui_solutions/
It also shows the problems of each UI framework.

I am currently developing my own approach to this, where several things will be different and the target range is indeed from embedded up to full scale applications. It is at a very early stage and I am still developing on the interface/infrastructure design. It has a different architecture than other frameworks, allowing to use the same code on a client but having different abilities on the seat side, but I think this will be an advantage. The communication between client and seat is also designed to be run over a bidirectional connection like a SPI interface etc.

**Reply by baconpaul** [reply]
Yeah it’s an interesting problem 

I think the three questions are interesting: can a blind user use your ui, can a Japanese user use it, and can it render the preset name “synth 🌊 for 🐰” help define the difference between a frame buffer and a toolkit (as you know, defiantnerd, because we’ve discussed this). None of the commonly used ui frameworks used in plugins can do all three today.

**Reply by defiantnerd** [reply]
Indeed. You nailed it @baconpaul: accessibility and full utf8 support would be king and I think it is possible to cover them.

**Reply by baconpaul** [reply]
Oh and to your question upstream the clap-saw-demo at https://github.com/surge-synthesizer/clap-saw-demo is a standalone clap only MIT licensed synth which uses VSTGUI not QT to make a UI. There's an initial port of that to imgui also but it's relatively display driver independent. 

And if you want something even simpler https://github.com/baconpaul/clap-c99-distortion is a no-ui C99 implementation of a (really bad) distortion plugin which doesn't even use the helpers and so on.

**Comment by Photosounder** [comment]
Just to let you guys know that I did reach the first significant and exciting step of getting this done which is making a WASM "plugin" that updates a framebuffer provided by the host (the simple C file in the screenshot below, compiled to WASM by Emscripten's emcc, the resulting WASM "plugin" is only 276 bytes), then made a "host" that loads the WASM plugin, calls the expected function to update a 512x512 framebuffer every frame and then displays that framebuffer. From here I should be able to move quickly.

![2023-03-29 22 31 09](https://user-images.githubusercontent.com/7466235/228660435-9c4cd2e8-36bb-45f0-9279-5946f31e5ad9.png)

**Reply by baconpaul** [reply]
Did you know that the team deploying Cardinal (which is basically a fork of VCV Rack) can run the entire thing in wasm, including the entire surge engine and the rest of the DSP.

**Reply by Photosounder** [reply]
I never of any of those things before tbh. Does it run WASM plugins from a non-WASM host?

**Reply by baconpaul** [reply]
VCV Rack is an entire euro rack system which has a mixed open source  / closed source strategy.

Cardinal is a fork of the open source side which compiles a subset of the plugin suite statically but in doing so allows it to run in browser

My only point was: technologies like nanovg, audio pipelines, and reasonably complicated synthesis are already running in browser under WASM today and I thought that might help you to know. But it runs in browser; it doesn't embed a WASM engine into bitwig or ardour or anything. VCV Rack is (in one configuration) a standalone musical environment.

I have not had a lot of experience with this WASM deployment other than merging a few PRs which were needed to get surge running under WASM

**Reply by Photosounder** [reply]
Cool, I've been making JavaScript/WASM builds of a couple of my C programs since 2017 with emscripten, but making and running WASM "plugins" from a C host is the new and exciting thing for me. Btw I wanted to make a standalone emscripten version of one of my plugins (a visualiser) but dealing with the JavaScript side (and AudioWorkletNode, which everyone seems to shun despite being the way to go for audio callbacks that aren't on the main thread) is really difficult and frustrating. Now I'm contemplating the much brighter perspective of making a bunch of compile once-run everywhere plugins that will run in native DAWs (at least through a bridge, but the bridge once stable won't need to be updated much).

Btw that whole concept is actually a lot like console emulation, you can look at old console ROMs as programs that can be interpreted on every platform the same way forever. Essentially I'm trying to make plugin "ROMs" that can be emulated anywhere forever.

**Reply by baconpaul** [reply]
yeah have you looked at what (say) jbridge did to make 32 bit plugins run in 64 bit hosts? or bluecat or kushview elements did for host-in-host for vst3 in pro tools and stuff? very different in some ways but not totally alien.

**Reply by Photosounder** [reply]
I'm aware of those from recommending them to my users, but I haven't used them. I probably should, I have questions about how they work, well, mainly about how parameter automation works.

**Reply by abique** [reply]
How do you plan to get mouse events? Change the mouse cursor, control the position of the mouse cursor during a mouse interaction?

**Reply by abique** [reply]
Here is an idea:
- You'll need more than what clap provides: a full windowing system abstraction (and it's OK), because drag and drop, copy paste, shaders, ...
  - in a similar way to WASI, you'll provide a runtime that is complementary to WASI to provide the extra GUI/Desktop features
- We can bring this technology without modifying existing hosts, so it is accessible to all

Approach:
- create the wasm plugin's "support interfaces/runtime", this runtime has to be versioned and the plugin explicitly mention their min version requirement; version N+1 should be backward compatible with version N
- create a wasm bridge plugin which will proxy and host wasm clap plugin:
  - this plugin is linked with a wasm runtime, will do the 1:1 translation layer from wasm to native as well as providing all the support interfaces to the wasm plugin
  - the user installs this wasm bridge, and all its hosts will be then able to load wasm plugins
  - the wasm bridge plugin will take care of scanning all the wasm plugins and merge the factories together
- it could be possible to create a wasm host support library, basically only three public functions are needed: `clap_wasm_dlopen()`, `clap_wasm_dlsym()`, `clap_wasm_dlclose()`? But this should not be recommended because the user may not be able to update the version of the wasm support lib that is linked with the host

Conclusion:
- with this approach we can get wasm for free in every hosts which supports native clap plugins
  - the deployment and adoption of this tech from the host side is resolved
- we can have as much "GUI & Desktop" support library as we want in the plugin proxy
  - if this interface already includes a vector rendering, font rendering, etc... it is then much less work for the plugin manufacturer to jump on wasm, because I can predict that if you ask people to build freetype, harfbuzz, cairo, libpng, and all the minimum requirement to make descent interactive GUI then you'll loose 95% on the way.
  - we can start with very few interfaces, and add more step by step:
    1. create window and paint buffer
    2. get keyboard and mouse events
    3. get touch and pen events (touch screen)
    4. access desktop copy paste buffer
    5. drag and drop
    6. maybe add a vector rendering library (font rendering, image loading and rendering, cairo like?)
    7. etc...

**Reply by Photosounder** [reply]
> How do you plan to get mouse events? Change the mouse cursor, control the position of the mouse cursor during a mouse interaction?

For my IPlug2 bridge plan, just like I did with SplineEQ and Spiral (I have a good habit of avoiding relying on features specific to libraries so I did everything to make all such things be handled my own way, although not entirely, I still relied on invisible IPlug controls for knobs): have one control that represents the whole window, then in `IControl::OnMouseDown()` and the likes call the WASM function to receive mouse events. Everything is just about taking things from where you'd normally get them then passing them on to the WASM plugin, you can basically copy what you do with IPlug2. That goes for any similar question, if IPlug2 takes drag-and-drop then I can pass them on to the WASM plugin, if it doesn't, then it's an IPlug2 problem, but I'm confident that IPlug2 has everything truly necessary. Shaders aren't truly needed, I already addressed this, there's no truly portable and universal way to do this that I know of that will still work in 30 years, if you want shaders you want ephemeral platform-specific builds, how else would you currently have shaders on macOS? Of course I could offer a layer of abstraction for drawing commands of primitives, but that's not part of my initial goal.

One way to look at the initial lack of such features is that the IPlug2 host (or any other maintained host) will always be at least as recent as any WASM plugin, so if anything becomes implemented in the host, provided that it was done by someone wise enough to not implement something that will stop working in 10 years, then you can start taking that new feature for granted. So let's say in a few years WebGPU becomes the true universal eternal standard (doubtful), then I could add this as something a plugin could use and start taking for granted.

If I were to make a CLAP host then I guess I would go the low effort route of using SDL/rouziclib (my library that has a GUI system based on SDL and it does windows, raster graphics, OpenCL graphics, of course all the SDL stuff, keyboard, mouse, touch (though I haven't tried that myself), string input (I even made some pretty advanced text editor as a basic control type), drag-and-drop, clipboard, my library also has built-in image and audio file saving and loading and OS-specific file dialogs), but then you'd need to ship SDL along with this. Maybe that would be better for me than using IPlug2 actually. Btw I really don't know why you guys insist on talking about offering things like font rendering, image loading or vector rendering or any of that when you don't even offer a window and framebuffer, however you want to do those things is the plugin developer's problem and using WASM doesn't prevent them for doing whatever they want, like you want to load PNGs, just add stb_image to your WASM plugin, I don't see why anyone would want to depend on what the plugin framework offers for rendering when it will cause trouble if they want to ditch the framework and move on to something else.

Btw it looks like I'm gonna have to deal with many small technical problems, all that WASM stuff is still a bit rough around the edges. I'm already dealing with inconsistencies in whether or not host initialisation works depending on how the host is launched, the main SDK compiler produces invalid binaries, and more incredibly, in a move that deserves an award for idiocy in design, dynamic allocation on WASM very often returns 0 pointers which are valid buffers, as in you can have a very valid buffer that you can use but is effectively NULL, and there's no clear way to get around this that works... 😂 But at least I was able to pass a dynamically allocated string in a call to a host function, so that's good, now I can call functions and pass data both ways.

**Reply by abique** [reply]
To continue with my proposal above, what if this wasm plugin bridge was providing a full html browser context, so you can access webgl, canvas, mouse events, etc...?
This bridge would be a bit heavy but probably lighter than your web browser, and you'd not have to worry about designing interfaces and maintaining backward compat

**Reply by Photosounder** [reply]
> To continue with my proposal above, what if this wasm plugin bridge was providing a full html browser context, so you can access webgl, canvas, mouse events, etc...? This bridge would be a bit heavy but probably lighter than your web browser, and you'd not have to worry about designing interfaces and maintaining backward compat

An intriguing idea, but I'm not sure how that would work, like, how would the WASM plugin send graphical data, in what form. You're free to try of course.
